extensions/smb/SmbConnectionControllerService.cpp

```diff
@@ -0,0 +1,107 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "SmbConnectionControllerService.h"
+#include "core/Resource.h"
+#include "utils/OsUtils.h"
+#include "utils/expected.h"
+
+namespace org::apache::nifi::minifi::extensions::smb {
+
+void SmbConnectionControllerService::initialize() {
+  setSupportedProperties(Properties);
+}
+
+void SmbConnectionControllerService::onEnable()  {
+  std::string hostname;
+  std::string share;
+
+  if (!getProperty(Hostname, hostname))
+    throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Missing hostname");
+
+  if (!getProperty(Share, share))
+    throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Missing share");
+
+
+  server_path_ = "\\\\" + hostname + "\\" + share;
+
+  auto password = getProperty(Password);
+  auto username = getProperty(Username);
+
+  if (password.has_value() != username.has_value())
+    throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Either a username and password must be provided, or neither of them should be provided.");
+
+  if (username.has_value())
+    credentials_.emplace(Credentials{.username = *username, .password = *password});
+  else
+    credentials_.reset();
+
+  ZeroMemory(&net_resource_, sizeof(net_resource_));
+  net_resource_.dwType = RESOURCETYPE_DISK;
+  net_resource_.lpLocalName = nullptr;
+  net_resource_.lpRemoteName = server_path_.data();
+  net_resource_.lpProvider = nullptr;
+}
+
+void SmbConnectionControllerService::notifyStop() {
+  auto disconnection_result = disconnect();
+  if (!disconnection_result)
+    logger_->log_error("Error while disconnecting from SMB: %s", disconnection_result.error().message());
+}
+
+nonstd::expected<void, std::error_code> SmbConnectionControllerService::connect() {
+  auto connection_result = WNetAddConnection2A(&net_resource_,
+      credentials_ ? credentials_->password.c_str() : nullptr,
+      credentials_ ? credentials_->username.c_str() : nullptr,
+      CONNECT_TEMPORARY);
+  if (connection_result == NO_ERROR)
+    return {};
+
+  return nonstd::make_unexpected(utils::OsUtils::windowsErrorToErrorCode(connection_result));
+}
+
+nonstd::expected<void, std::error_code> SmbConnectionControllerService::disconnect() {
+  auto disconnection_result = WNetCancelConnection2A(server_path_.c_str(), 0, true);
+  if (disconnection_result == NO_ERROR)
+    return {};
+
+  return nonstd::make_unexpected(utils::OsUtils::windowsErrorToErrorCode(disconnection_result));
+}
+
+bool SmbConnectionControllerService::isConnected() {
+  std::error_code error_code;
+  auto exists = std::filesystem::exists(server_path_, error_code);
+  if (error_code)
+    return false;
```

Comment:
Added some debug logging in https://github.com/apache/nifi-minifi-cpp/pull/1634/commits/256dca6a32fbd67737f87af2517c0d98fdbd3236#diff-8f6654a3a37c2021e4cbe35a2c93791afb67776b83389a56a45bf78e3f32bc3eR101 (didnt want to warn or error because we if the consequent connection fails aswell we will log an error)